libtenzir/src/execution_node.cpp

```diff
@@ -762,21 +758,18 @@ struct exec_node_state {
     TENZIR_TRACE("{} {} schedules run with a delay of {}", *self, op->name(),
                  data{backoff});
     run_scheduled = true;
+    auto async_run = [this] {
+      auto time_scheduled_guard = make_timer_guard(metrics.time_scheduled);
+      run();
+    };
     if (backoff == duration::zero()) {
-      self->mail(atom::internal_v, atom::run_v).send(self);
+      self->schedule_fn(std::move(async_run));
```

Comment:
Removing the indirection over a message here causes an infinite hang for source operators that never suspend themselves via `ctrl.set_waiting(false)`. This happens because CAF always prefers scheduled actions over messages, so this causes messages to never be read for such operators.

I've asked upstream about a fix for this.